stbt/android.py

```diff
@@ -0,0 +1,406 @@
+
+        """
+        x1, y1 = _region_to_tuple(start_position)
+        x2, y2 = _region_to_tuple(end_position)
+        stbt.debug("AdbDevice.swipe((%d,%d), (%d,%d))" % (x1, y1, x2, y2))
+
+        x1, y1 = self._to_native_coordinates(x1, y1)
+        x2, y2 = self._to_native_coordinates(x2, y2)
+        command = ["shell", "input", "swipe",
+                   str(x1), str(y1), str(x2), str(y2)]
+        self.adb(command, timeout_secs=10)
+
+    def tap(self, position):
+        """Tap on a particular location.
+
+        :param position: A `stbt.Region`, or an (x,y) tuple.
+
+        Example::
+
+            tap((100, 20))
+            tap(stbt.match(...).region)
+
+        """
+        x, y = _region_to_tuple(position)
+        stbt.debug("AdbDevice.tap((%d,%d))" % (x, y))
+
+        x, y = self._to_native_coordinates(x, y)
+        self.adb(["shell", "input", "tap", str(x), str(y)], timeout_secs=10)
+
+    def _adb(self, command, timeout_secs=None, **kwargs):
+        _command = []
+        if timeout_secs is not None:
+            _command += ["timeout", "%fs" % timeout_secs]
+        _command += [self.adb_binary]
+        if self.adb_server:
+            _command += ["-H", self.adb_server]
+        if self.adb_device:
+            _command += ["-s", self.adb_device]
+        _command += command
+        stbt.debug("AdbDevice.adb: About to run command: %r\n" % _command)
+        output = subprocess.check_output(
+            _command, stderr=subprocess.STDOUT, **kwargs)
+        if command[0] in ("connect", "tcpip", "usb"):
+            time.sleep(5)
+        return output
+
+    def _connect(self, timeout_secs):
+        if not self.adb_device:
+            raise RuntimeError('AdbDevice: error: If "tcpip=True" '
+                               'you must specify "adb_device"')
+        try:
+            if self.adb_device in self._adb(["devices"]):
+                return
+        except subprocess.CalledProcessError:
+            pass
+
+        # "adb connect" always returns success; we have to parse the output
+        # which looks like "connected to 192.168.2.163:5555" or
+        # "already connected to 192.168.2.163:5555" or
+        # "unable to connect to 192.168.2.100:5555".
+        output = self._adb(["connect", self.adb_device], timeout_secs)
+        if ("connected to %s" % self.adb_device) not in output:
+            sys.stderr.write(output)
+            raise AdbError(0, "adb connect %s" % self.adb_device, output, self)
+
+    def _to_native_coordinates(self, x, y):
+        if self.coordinate_system == CoordinateSystem.ADB_NATIVE:
+            return x, y
+        else:
+            return _to_native_coordinates(
+                x, y, self.coordinate_system, self._get_display_dimensions())
+
+    def _get_display_dimensions(self):
+        return _parse_display_dimensions(
+            self.adb(["shell", "dumpsys", "window"],
+                     timeout_secs=10, capture_output=True))
+
+
+class AdbError(Exception):
```

Comment:
Not sure. I think of subprocess as an implementation detail and I have considered changing the `__str__` to be less CalledProcessError-like, and possibly changing `cmd` to only show the adb command not the full shell command.